src/computability/tm_computable.lean

```diff
@@ -0,0 +1,229 @@
+                (λ h, begin rw h, exact s, end)
+                (λ _,[]) }
+
+/-- The final configuration corresponding to a list in the output alphabet. -/
+def halt_list (tm : fin_tm2) (s : list (tm.Γ tm.k₁)) : tm.cfg :=
+{ l := option.none,
+  var := tm.initial_state,
+  stk := λ k, @dite (k = tm.k₁) (tm.K_decidable_eq k tm.k₁) (list (tm.Γ k))
+                (λ h, begin rw h, exact s, end)
+                (λ _,[]) }
+
+/-- A "proof" of the fact that f eventually reaches b when repeatedly evaluated on a,
+remembering the number of steps it takes. -/
+structure evals_to {σ : Type*} (f : σ → option σ) (a : σ) (b : option σ) :=
+(steps : ℕ)
+(evals_in_steps : ((flip bind f)^[steps] a) = b)
+
+/-- A "proof" of the fact that f eventually reaches b in at most m steps when repeatedly evaluated on a,
+remembering the number of steps it takes. -/
+structure evals_to_in_time {σ : Type*} (f : σ → option σ) (a : σ) (b : option σ) (m : ℕ) extends evals_to f a b :=
+(steps_le_m : steps ≤ m)
+
+/-- Reflexivity of `evals_to` in 0 steps. -/
+@[refl] def evals_to.refl {σ : Type*} (f : σ → option σ) (a : σ) : evals_to f a a := ⟨0,rfl⟩
+
+/-- Transitivity of `evals_to` in the sum of the numbers of steps. -/
+@[trans] def evals_to.trans {σ : Type*} (f : σ → option σ) (a : σ) (b : σ) (c : option σ)
+  (h₁ : evals_to f a b) (h₂ : evals_to f b c) : evals_to f a c :=
+⟨h₂.steps + h₁.steps,
+ by rw [function.iterate_add_apply,h₁.evals_in_steps,h₂.evals_in_steps]⟩
+
+/-- Reflexivity of `evals_to_in_time` in 0 steps. -/
+@[refl] def evals_to_in_time.refl {σ : Type*} (f : σ → option σ) (a : σ) : evals_to_in_time f a a 0 :=
+⟨evals_to.refl f a, le_refl 0⟩
+
+/-- Transitivity of `evals_to_in_time` in the sum of the numbers of steps. -/
+@[trans] def evals_to_in_time.trans {σ : Type*} (f : σ → option σ) (a : σ) (b : σ) (c : option σ)
+  (m₁ : ℕ) (m₂ : ℕ) (h₁ : evals_to_in_time f a b m₁) (h₂ : evals_to_in_time f b c m₂) :
+  evals_to_in_time f a c (m₂ + m₁) :=
+⟨evals_to.trans f a b c h₁.to_evals_to h₂.to_evals_to, add_le_add h₂.steps_le_m h₁.steps_le_m⟩
+
+/-- A proof of tm outputting l' when given l. -/
+def tm2_outputs (tm : fin_tm2) (l : list (tm.Γ tm.k₀)) (l' : option (list (tm.Γ tm.k₁))) :=
+evals_to tm.step (init_list tm l) ((option.map (halt_list tm)) l')
+
+/-- A proof of tm outputting l' when given l in at most m steps. -/
+def tm2_outputs_in_time (tm : fin_tm2) (l : list (tm.Γ tm.k₀)) (l' : option (list (tm.Γ tm.k₁))) (m : ℕ) :=
+evals_to_in_time tm.step (init_list tm l) ((option.map (halt_list tm)) l') m
+
+/-- The forgetful map, forgetting the upper bound on the number of steps. -/
+def tm2_outputs_in_time.to_tm2_outputs {tm : fin_tm2} {l : list (tm.Γ tm.k₀)}
+{l' : option (list (tm.Γ tm.k₁))} {m : ℕ} (h : tm2_outputs_in_time tm l l' m) : tm2_outputs tm l l' :=
+h.to_evals_to
+
+/-- A Turing machine with input alphabet equivalent to Γ₀ and output alphabat equivalent to Γ₁. -/
+private structure computable_by_tm2_aux (Γ₀ Γ₁ : Type) :=
```

Comment:
```suggestion
/-- A Turing machine with input alphabet equivalent to Γ₀ and output alphabet equivalent to Γ₁. -/
structure computable_by_tm2_aux (Γ₀ Γ₁ : Type) :=
```
Don't use private ever. You will regret it later. An `_aux` like naming scheme is sufficient to indicate "implementation detail", but you never know when someone will want to prove something more about your implementation details.